ding/torch_utils/network/gtrxl.py

```diff
@@ -0,0 +1,536 @@
+    """
+    def __init__(
+        self,
+        input_dim: int,
+        head_dim: int = 128,
+        embedding_dim: int = 256,
+        head_num: int = 2,
+        mlp_num: int = 2,
+        layer_num: int = 3,
+        memory_len: int = 64,
+        dropout_ratio: float = 0.,
+        activation: nn.Module = nn.ReLU(),
+    ) -> None:
+        """Overview:
+            Init GTrXL Model
+        Arguments:
+            - input_dim (:obj:`int`): dimension of input (dimension of a single observation)
+            - head_dim (:obj:`int`): dimension of each head
+            - hidden_dim (:obj:`int`): dimension of hidden layer in mlp
+            - embedding_dim (:obj:`int`): dimension of embedding (dimension of a single observation after embedding)
+            - head_num (:obj:`int`): number of heads for multihead attention
+            - mlp_num (:obj:`int`): number of mlp layers in attention layer
+            - layer_num (:obj:`int`): number of transformer layers
+            - dropout_ratio (:obj:`float`): dropout ratio
+            - activation (:obj:`nn.Module`): activation function
+        """
+        super(GTrXL, self).__init__()
+        assert embedding_dim % 2 == 0, 'embedding_dim={} should be even'.format(input_dim)
+        self.head_num = head_num
+        self.head_dim = head_dim
+        self.layer_num = layer_num
+        self.embedding = fc_block(input_dim, embedding_dim, activation=activation)
+        self.activation = activation
+        self.pos_embedding = PositionalEmbedding(embedding_dim)
+        # memory to save hidden states of past segments
+        # it will be initialized in the forward method to get its size dynamically
+        self.memory = None
+        self.memory_len = memory_len
+        layers = []
+        dims = [embedding_dim] + [embedding_dim] * layer_num
+        self.dropout = nn.Dropout(dropout_ratio)
+        for i in range(layer_num):
+            layers.append(
+                GatedTransformerXLLayer(dims[i], head_dim, embedding_dim, head_num, mlp_num, self.dropout,
+                                        self.activation)
+            )
+        self.layers = nn.Sequential(*layers)
+        self.embedding_dim = embedding_dim
+        # u and v are the parameters to compute global content bias and global positional bias
+        self.u, self.v = (
+            torch.nn.Parameter(torch.Tensor(self.head_num, self.head_dim)),
+            torch.nn.Parameter(torch.Tensor(self.head_num, self.head_dim)),
+        )
+
+    def reset(self, batch_size: int = None, state: Optional[torch.Tensor] = None):
+        r"""
+        Overview:
+            Clear or set the memory of GTrXL.
+         Arguments:
+            - batch_size (:obj:`int`): batch size
+            - state (:obj:`Optional[torch.Tensor]`): input memory.
+            Shape is (layer_num, memory_len, bs, embedding_dim).
+        """
+        self.memory = Memory()
+        if batch_size is not None:
```

Comment:
We need to check batch size because it could happen that an input x has a different batch size than the batch size of the existing memory m. This could happen when we switch from collect to train phase for example. In this case the memory will be reset. In a next version, each phase could have its own memory with a wrapper so to avoid this problem.